core/shared/src/main/scala/zio/stm/TArray.scala

```diff
@@ -16,139 +16,193 @@
-    new TArray(array.reverse).findM(predicate)
+  /**
+   * Find the last element in the array matching an STM predicate.
+   */
+  final def findLastM[E](p: A => STM[E, Boolean]): STM[E, Option[A]] =
+    new TArray(array.reverse).findM(p)
 
-  /** Find the first element in the array matching an STM predicate. */
-  final def findM[E](predicate: A => STM[E, Boolean]): STM[E, Option[A]] =
+  /**
+   * Find the first element in the array matching an STM predicate.
+   */
+  final def findM[E](p: A => STM[E, Boolean]): STM[E, Option[A]] =
     if (array.isEmpty) STM.succeed(None)
     else
       array.head.get.flatMap { a =>
-        predicate(a).flatMap { result =>
+        p(a).flatMap { result =>
           if (result) STM.succeed(Some(a))
-          else new TArray(array.tail).findM(predicate)
+          else new TArray(array.tail).findM(p)
         }
       }
 
-  /** The first entry of the array, if it exists. */
-  final def firstOption: STM[Nothing, Option[A]] = if (array.isEmpty) STM.succeed(None) else array.head.get.map(Some(_))
+  /**
+   * The first entry of the array, if it exists.
+   */
+  final def firstOption: STM[Nothing, Option[A]] =
+    if (array.isEmpty) STM.succeed(None) else array.head.get.map(Some(_))
 
-  /** Atomically folds [[TArray]] with pure function. */
+  /**
+   * Atomically folds [[TArray]] with pure function.
+   */
   final def fold[Z](acc: Z)(op: (Z, A) => Z): STM[Nothing, Z] =
     if (array.isEmpty) STM.succeed(acc)
     else array.head.get.flatMap(a => new TArray(array.tail).fold(op(acc, a))(op))
 
-  /** Atomically folds [[TArray]] with STM function. */
+  /**
+   * Atomically folds [[TArray]] with STM function.
+   */
   final def foldM[E, Z](acc: Z)(op: (Z, A) => STM[E, Z]): STM[E, Z] =
     if (array.isEmpty) STM.succeed(acc)
     else
       array.head.get.flatMap { a =>
         op(acc, a).flatMap(acc2 => new TArray(array.tail).foldM(acc2)(op))
       }
 
-  /** Atomically evaluate the conjunction of a predicate across the members of the array. */
-  final def forall(predicate: A => Boolean): STM[Nothing, Boolean] = exists(a => !predicate(a)).map(!_)
+  /**
+   * Atomically evaluate the conjunction of a predicate across the members of the array.
+   */
+  final def forall(p: A => Boolean): STM[Nothing, Boolean] = exists(a => !p(a)).map(!_)
 
-  /** Atomically evaluate the conjunction of an STM predicate across the members of the array. */
-  final def forallM[E](predicate: A => STM[E, Boolean]): STM[E, Boolean] =
-    countM(predicate).map(_ == array.length) //Universal quantification should not be order dependent so the entire array needs to be observed.
+  /**
+   * Atomically evaluate the conjunction of an STM predicate across the members of the array.
+   */
+  final def forallM[E](p: A => STM[E, Boolean]): STM[E, Boolean] =
+    countM(p).map(_ == array.length)
 
-  /** Atomically performs side-effect for each item in array */
+  /**
+   * Atomically performs side-effect for each item in array.
```

Comment:
perform a transaction for each item in the array?

Reply:
Slightly different, but done.